### DISCUSSION TITLE
how to get new direct message in real-time?

### DISCUSSION THREAD

**Discussion by thaziz** [plhery/node-twitter-api-v2#179] [Q&A]
I want to be able to receive direct messages in real time

**Comment by r-ndom-developer** [comment]
I'm trying to implement this as well. It looks like there's a custom `getStream` method but none of the DM endpoints I know seem to work, they return a 404.

```typescript

const stream = await client.v1.stream.getStream('direct_message/events/new.json');

stream.on(
   ETwitterStreamEvent.Data,
   (data: any) => console.log(data)
);
  
stream.on(
   ETwitterStreamEvent.ConnectionError,
   (err: any) => console.log(err)
);
  
stream.on(
   ETwitterStreamEvent.ConnectionClosed,
   (err: any) => console.log(err)
);

Response:

{
	"type": "response",
	"code": 404,
	"error": "",
	"headers": {
		"date": "Sat, 12 Feb 2022 14:48:03 UTC",
		"server": "tsa_b",
		"set-cookie": [
			...
		],
		"cache-control": "no-cache, no-store, max-age=0",
		"content-length": "0",
		"strict-transport-security": "max-age=631138519",
		"x-response-time": "7",
		"x-connection-hash": "",
		"connection": "close"
	},
        rateLimit: undefined,
        data: ''
}
```

Not sure if the library supports this..

**Comment by alkihis** [comment]
Shortly:
There is no way to stream direct messages on public API now. Years ago, one stream method existed but it has been closed.

Live-time direct messages are available through Account Activity API, a paid API that runs with webhooks (so, cannot be used with this client, as it requires a HTTP server on your side).

**Reply by olsiodobashi** [reply]
I've used Twit for this with Autohook and something called ngrok and it was working just a few weeks ago. Is there no way to implement something similar using this library? Thanks

**Reply by alkihis** [reply]
Neither autohook or ngrok are linked to a library. They run a web server that listen to account activity API webhooks. You are free to use them and make "classic" HTTP requests to Twitter endpoints with this library.

**Reply by waptik** [reply]
@olsiodobashi: As @alkihis said, your best bet is to use webhooks to listen to Twitter account activity API.
Here are some examples [Tweets](https://github.com/ngVenezuela/wengy-ven/blob/develop/api/twitter/update.ts) & [DMs](https://github.com/addissemagn/api.addis.world/blob/master/src/controllers/twitter.js). It's pretty easy, though.

Oh, I forgot to mention the [Account Activity API](https://developer.twitter.com/en/docs/twitter-api/premium/account-activity-api/overview) is a paid API; you can use it for free as they have a free tier.
